packages/babel-helper-builder-react-jsx-experimental/src/index.js

```diff
@@ -0,0 +1,844 @@
+            "jsxStaticIdentifier",
+            createIdentifierParser(
+              createIdentifierName(
+                path,
+                options.development ? "jsxDEV" : "jsxs",
+                importName,
+              ),
+            ),
+          );
+
+          state.set(
+            "createElementIdentifier",
+            createIdentifierParser(
+              createIdentifierName(path, "createElement", importName),
+            ),
+          );
+
+          state.set(
+            "jsxFragIdentifier",
+            createIdentifierParser(
+              createIdentifierName(path, "Fragment", importName),
+            ),
+          );
+        } else {
+          throw path.buildCodeFrameError(
+            `Runtime must be either "classic" or "automatic".`,
+          );
+        }
+      }
+    },
+
+    exit(path, state) {
+      if (
+        state.get("runtime") === "classic" &&
+        state.get("pragmaSet") &&
+        state.get("usedFragment") &&
+        !state.get("pragmaFragSet")
+      ) {
+        throw new Error(
+          "transform-react-jsx: pragma has been set but " +
+            "pragmaFrag has not been set",
+        );
+      }
+    },
+  };
+
+  return visitor;
+
+  // We want to use React.createElement, even in the case of
+  // jsx, for <div {...props} key={key} /> to distinguish it
+  // from <div key={key} {...props} />. This is an intermediary
+  // step while we deprecate key spread from props. Afterwards,
+  // we will stop using createElement in the transform.
+  function shouldUseCreateElement(path) {
+    const openingPath = path.get("openingElement");
+    const attributes = openingPath.node.attributes;
+
+    let seenPropsSpread = false;
+    for (let i = 0; i < attributes.length; i++) {
+      const attr = attributes[i];
+      if (
+        seenPropsSpread &&
+        t.isJSXAttribute(attr) &&
+        attr.name.name === "key"
+      ) {
+        return true;
+      } else if (t.isJSXSpreadAttribute(attr)) {
+        seenPropsSpread = true;
+      }
+    }
+    return false;
+  }
+
+  function createIdentifierName(path, name, importName) {
+    if (isModule(path) && importName) {
+      const identifierName = `${importName[name]}`;
+      return identifierName;
+    } else {
+      return `${importName}.${name}`;
+    }
+  }
+
+  function getImportNames(parentPath) {
+    const imports = {};
+
+    parentPath.traverse({
+      JSXElement(path) {
+        const openingPath = path.get("openingElement");
+        openingPath.parent.children = t.react.buildChildren(openingPath.parent);
+
+        if (shouldUseCreateElement(path)) {
+          imports.createElement = true;
+        } else if (path.node.children.length > 1) {
+          const importName = options.development ? "jsxDEV" : "jsxs";
+          imports[importName] = true;
+        } else {
+          const importName = options.development ? "jsxDEV" : "jsx";
+          imports[importName] = true;
+        }
+      },
```

Comment:
If we have already found all the possible import names, can we call `path.stop()` to speed up the traversal?